### DISCUSSION TITLE
OAuth2 support

### DISCUSSION THREAD

**Discussion by StijnCaerts** [stac-utils/stac-asset#143] [Ideas]
Hi, we're interested in adding OAuth2/OIDC support to the HttpClient. Would you like this to be a configuration option on the existing `HttpClient`, or should we create a dedicated class for this? We can use [the `aiohttp-oauth2-session` package](https://pypi.org/project/aiohttp-oauth2-session/) to add OAuth2 token integration to the `aiohttp` `ClientSession` object.
It would also be nice if the configuration can be automatically loaded from the _authentication_ STAC extension, but that would probably require some changes to the way clients are loaded: https://github.com/stac-utils/stac-asset/blob/a201bba6e04e60494f77fffa73c6260f422d1505/src/stac_asset/client.py#L189-L221

What are your thoughts on this?

**Comment by gadomski** [comment]
Thanks for the interest and the idea! At first glance, I _think_ it makes sense to add to the existing `HttpClient`, and we can break it out later if it feels awkward. And 👍🏼 to automatically detecting the extension, that would be super-cool.

**Comment by StijnCaerts** [comment]
I'm looking into this feature again, but I was wondering what the best way is to handle all the configuration parameters.
Things like:

- OAuth2 grant type
- token / authorization URL
- client ID
- username, password
- ...

Should these all go in the `Config` dataclass?

Here is a small example of how it can be used with an external dependency:
```python
import stac_asset
from aiohttp_oauth2_client.client import OAuth2Client
from aiohttp_oauth2_client.grant.device_code import DeviceCodeGrant

...

async with DeviceCodeGrant(
    token_url=TOKEN_URL,
    device_authorization_url=DEVICE_AUTH_URL,
    client_id=CLIENT_ID,
    pkce=True,
) as grant, OAuth2Client(grant) as client:
    download_client = stac_asset.HttpClient(session=client)
    await stac_asset.download_item_collection(item_collection, "./Downloads", clients=[download_client])
```

**Reply by gadomski** [reply]
I _think_ I'm 👎🏼 on putting them in `Config`, at least for starters. As mentioned [here](https://github.com/stac-utils/stac-asset/pull/163#pullrequestreview-2056120364), I'm a little queasy about how some clients bubble their configuration up to `Config`, but not all/everything. I've opened https://github.com/stac-utils/stac-asset/issues/176 to document my unhappiness and one possible pseudo-solution.

All this to say, I'd build out your custom client first, and we can figure out how to wire it into `Config` and the CLI later.

**Comment by gadomski** [comment]
This has been implemented in https://github.com/stac-utils/stac-asset/discussions/143
